src/lib/operators/join_sort_merge.cpp

```diff
@@ -910,31 +891,74 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
 
     _perform_join();
 
-    // merge the pos lists into single pos lists
-    auto output_left = _concatenate_pos_lists(_output_pos_lists_left);
-    auto output_right = _concatenate_pos_lists(_output_pos_lists_right);
+    if (include_null_left || include_null_right) {
+      auto null_output_left = std::make_shared<PosList>();
+      auto null_output_right = std::make_shared<PosList>();
 
-    // Add the outer join rows which had a null value in their join column
-    if (include_null_left) {
-      for (auto row_id_left : *_null_rows_left) {
-        output_left->push_back(row_id_left);
-        output_right->push_back(NULL_ROW_ID);
+      // Add the outer join rows which had a null value in their join column
+      if (include_null_left) {
+        null_output_left->reserve(_null_rows_left->size());
+        null_output_right->insert(null_output_right->end(), _null_rows_left->size(), NULL_ROW_ID);
+        for (const auto& row_id_left : *_null_rows_left) {
+          null_output_left->push_back(row_id_left);
+        }
+      }
+      if (include_null_right) {
+        null_output_left->insert(null_output_left->end(), _null_rows_right->size(), NULL_ROW_ID);
+        null_output_right->reserve(_null_rows_right->size());
+        for (const auto& row_id_right : *_null_rows_right) {
+          null_output_right->push_back(row_id_right);
+        }
+      }
+
+      DebugAssert(null_output_left->size() == null_output_right->size(),
+                  "Null positions lists are expected to be of equal length.");
+      if (!null_output_left->empty()) {
+        _output_pos_lists_left.push_back(null_output_left);
+        _output_pos_lists_right.push_back(null_output_right);
       }
     }
-    if (include_null_right) {
-      for (auto row_id_right : *_null_rows_right) {
-        output_left->push_back(NULL_ROW_ID);
-        output_right->push_back(row_id_right);
+
+    // Intermediate structure for output chunks (to avoid concurrent appending to table)
+    std::vector<std::shared_ptr<Chunk>> output_chunks(_output_pos_lists_left.size());
+
+    // Determine if writing output in parallel is necessary.
+    // As partitions ought to be roughly equally sized, looking at the first should be sufficient.
+    const auto write_output_concurrently = _cluster_count > 1 && _output_pos_lists_left[0]->size() > 10'000;
+
+    std::vector<std::shared_ptr<AbstractTask>> output_jobs;
+    output_jobs.reserve(_output_pos_lists_left.size());
```

Comment:
i don't know how much performance impact it is, but couldn't we avoid reserving extra memory in case we don't write concurrently in the end?

Reply:
The number of output lists is given by the radix count. For small joins, where the `reserve()` could be noticable, the number of pos lists is very small and it should basically come for free.
In other cases, the joins are to large that it will never be noticable anyway.

It's probably impossible to measure the impact given the general complexity of join operators, but I'd vote for reserving vector whenever we have a decent idea how large they might become.